src/Terminal/commands/grep.ts

```diff
@@ -0,0 +1,179 @@
+import { Terminal } from "../../Terminal";
+import { BaseServer } from "../../Server/BaseServer";
+import { hasTextExtension } from "../../Paths/TextFilePath";
+import { ContentFile, ContentFilePath, allContentFiles } from "../../Paths/ContentFile";
+import { FilePath } from "src/Paths/FilePath";
+
+type LineParser = (line: string, i: number) => string;
+type PtLineParser = (filename: string, line: string, i: number) => string;
+type FileParser = (file: ContentFile | null) => string[] | string;
+type FileTuple = [ContentFilePath, ContentFile];
+
+interface OkArgs {
+  lineNum: string[];
+  regExpr: string[];
+  yesName: string[];
+  notName: string[];
+  toFile: string[];
+  overWrt: string[];
+  quiet: string[];
+}
+
+interface Options {
+  lineNum: boolean;
+  regExpr: boolean;
+  yesName: boolean;
+  notName: boolean;
+  toFile: boolean;
+  overWrt: boolean;
+  quiet: boolean;
+
+  multiFile: boolean;
+}
+
+const OK_ARGS: OkArgs = {
+  lineNum: ["-n", "--line-number"],
+  regExpr: ["-G", "--basic-regexp"],
+  yesName: ["-H", "--with-filename"],
+  notName: ["-h", "--no-filename"],
+  toFile: ["-O", "--output-number"],
+  overWrt: ["-f", "--allow-overwite"],
+  quiet: ["-q", "--quiet", "--silent"],
+};
+
+function getParseFunc(pattern: string | RegExp, options: Options, filename: string, line: string, i: number): string {
+  const RED: string = "\x1b[31m";
+  const DEF: string = "\x1b[0m"; // default
+  const GREEN: string = "\x1b[32m";
+  const CYAN: string = "\x1b[36m";
+  const MAGENTA: string = "\x1b[35m";
+
+  const editedLine: string = line.replaceAll(pattern, `${RED}$&${DEF}`);
+  if (line === editedLine) return ""; // don't print unmatched lines
+  if (options.toFile) return line; // don't colour when outputting to file
+  const name: string =
+    (options.multiFile || options.yesName) && !options.notName ? `${MAGENTA}${filename}${CYAN}:${DEF}` : "";
+  const lineNo: string = options.lineNum ? `${GREEN}${i + 1}${CYAN}:${DEF}` : "";
+  const prefix: string = name + lineNo;
+
+  return prefix + editedLine;
+}
+
+function parseFile(parseFunc: PtLineParser): FileParser {
+  return function (file: ContentFile | null): string | string[] {
+    if (!file) return "";
+    const content: string = "content" in file ? file["content"] : file["code"];
+
+    const parseLine: LineParser = parseFunc.bind(null, file.filename);
+
+    const editedContent: string | string[] = content.split("\n").map(parseLine);
```

Comment:
```suggestion
    const editedContent: string[] = [];
    for (const line of file.content.split("\n")) {
      const edited = parseFunc(file.filename, line);
      if (edited !== "") {
        editedContent.push(edited);
      }
    }
```
Optional: This might be a place where being iterative is worthwhile for speed.